### PR TITLE
331: handle failed fetching consistently on all services

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary

This PR mainly addresses:

1. Fix bug that wouldn't increment failed counter stats on product and category sync. 
2. Fix bug that would continue to attempt syncing in productType sync after a failed fetch.
3. Handle fail fetching and not found fetching during retry of concurrency modification exception in product and category sync.
3. Handeling errors on fetching in a consistent way across all different syncs/services/fetch points.
4. Decoupling services from sync statistics incrementing.
5. Fix potential NPE in `AttributeDefinitionCustomBuilder`.
6. Some boyscout changes.

#### Hints for Review

1. Decouple of handeling of error on fetch from services:
- [`ProductService`](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-6dfa47f1d198543a343a810f8a0c0ab5)
- [`CategoryService`](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-38b71d2c035833260351074420bf8067)

2. Adapt [integration tests for CategoryServiceImpl](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-d0865f28893ee5e72f220a041a594d8e) after changing of handeling of error on fetch.

3. Adapt [integration tests for ProductServiceImpl](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-99e884322fcfe598d4c632ce723ad4ca) after changing of handeling of error on fetch.

4. Handle fail on fetch on `cacheKeysToIds`, `fetchMatchingCategoriesByKeys` and `fetchCategory` (which is called while retrying for concurrency modification exception) in [CategorySync](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-44d2fa184e625bcc0900b105f60d8677).  Trigger error call back and increment failed stats.

5. Add [unit tests `CategorySync` for error on cache or fetch](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-b2dcd7d1b34c29a7bb3d0c9f77c8f69b)

6. Handle fail on fetch on `cacheKeysToIds`, `fetchMatchingProductsByKeys ` and `fetchProduct` (which is called while retrying for concurrency modification exception) in [ProductSync](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-0d16321dedd4645c5b1ad8e28f5aa44c). Trigger error call back and increment failed stats.

7. Add [unit tests `ProductSync` for error on cache or fetch](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-3c3599c7a6dbce2afca42e82dfe52549)

8. Fix [mocking of client throwing concurrent modification exception on update.](https://github.com/commercetools/commercetools-sync-java/blob/1dd5a7acbb784d08de33dd5cb97a7cc99216f529/src/integration-test/java/com/commercetools/sync/integration/externalsource/categories/CategorySyncIT.java#L210-L219) on CategorySync Integration Test. 

9. Add integration tests for handling concurrency modification exception retry, fail on fetch and and not found on fetch on the [CategorySyncIT](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-7a2564580078e482081eea430630ad3d).

10. Add integration tests for handling concurrency modification exception retry, fail on fetch and and not found on fetch on the [ProductSyncIT](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-d68ebb35acffe45efc007ae884c446b0).

11. [Change handeling of errors on `productType` fetching](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-61da62440dbd32893e8a01d615654679):

- Make `ProductTypeSync` constructor only accessible within the package for unit tests. 
- Fix handeling of error of fetch from `fetchMatchingProductTypesByKeys` (in the beginning it used to return an `emptySet` on error and continue trying to sync, but this is /incorrect behaviour. Now on error, it doesn't continue trying to create.
- Add unit test for [error on `fetchMatchingProductTypesByKeys`](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-cc855ec5233869e3522426b02b1a4c70).

##### Boy-scout changes

1. Handle potential NPEs on [AttributeDefinitionCustomBuilder](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-e4c816ad4b087d38a7a7d7c2267e0e78) and [add unit tests](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-745f07eed55d0c5adea0028a2e71cc14).

2. Make all service classes final. Consequently, use the [mockito configuration](https://github.com/commercetools/commercetools-sync-java/pull/340/files#diff-163d44d5fcac882c8e47a683cebf3bf6) to be able to mock final classes.

#### Relevant Issues
#331 

#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [x] Add Release Notes entry.

### CODE REVIEW

**Comment by heshamMassoud** [comment]
Hi @ahmetoz 

> we should use the same naming convention for using keyToIdCache in java docs - for example, is it an in-memory map or cached map?

The thing in the Javadoc of the service interface itself should not be very coupled from the service implementation, since there could be more than one implementation for the service (even one that maybe doesn't use this map) That's why I tried as much as possible to make it coupled with inner implementation.

> we are using keyToIdCache in services maybe we can consider making this hashmap as a data-structure in a different class, so in the future, we can track how we are using the key to id cache.

Can you elaborate more on this suggestion, as I am not entirely sure what you mean?


Regarding the other two notes - will address on them now ;)

**Comment by heshamMassoud** [comment]
> updating the java-doc for fetchMatchingProductTypesByKeys would be nice to make it consistent between services.

feadff9

> it is better to remove @Ignored annotation for fetchMatchingProductTypesByKeys_WithBadGateWayExceptionAlways_ShouldFail and update it.

eadc8bd

@ahmetoz Thanks for catching them

**Comment by ahmetoz** [comment]
> Can you elaborate more on this suggestion, as I am not entirely sure what you mean?

I've just tried to implement it as an extra class and this seems not good(looks like extra wrapping). But maybe we can write this functions to `BaseService` like below but I realized, it's not a big improvement. 😅 

``` java

    private void cacheId(String key, String id) {
        keyToIdCache.put(key, id);
    }

    private Optional<String> getId(String key) {
        return Optional.ofNullable(keyToIdCache.get(key));
    }

    private boolean isCached(String key){
        return keyToIdCache.containsKey(key);
    }

```

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncIT.java [hunk 1]
```diff
@@ -345,6 +346,132 @@ public void sync_withChangedProductButConcurrentModificationException_shouldRetr
...
+        // preparation
+        final SphereClient spyClient = buildClientWithConcurrentModificationUpdateAndFailedFetchOnRetry();
+
+        final ProductSyncOptions spyOptions = ProductSyncOptionsBuilder.of(spyClient)
+                                                                       .errorCallback(
+                                                                           (errorMessage, exception) -> {
+                                                                               errorCallBackMessages
+                                                                                   .add(errorMessage);
+                                                                               errorCallBackExceptions
+                                                                                   .add(exception);
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncIT.java]: thanks: fa2d2ff (just changed the name of the callback to what it's actually doing ;) )

File: src/main/java/com/commercetools/sync/categories/CategorySync.java [hunk 2]
```diff
@@ -161,27 +167,52 @@ public CategorySync(@Nonnull final CategorySyncOptions syncOptions) {
...
-                                                            updateCategoriesInParallel(categoryDraftsToUpdate))
-                                                        .thenApply((ignoredResult) -> {
-                                                            statistics.incrementProcessed(numberOfNewDraftsToProcess);
-                                                            return statistics;
-                                                        });
-                              });
+        return categoryService
+                .cacheKeysToIds()
+                .handle(ImmutablePair::new)
+                .thenCompose(cachingResponse -> {
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: I agree this lambda got huge now. I split it into smaller methods.

File: src/main/java/com/commercetools/sync/categories/CategorySync.java [hunk 3]
```diff
@@ -161,27 +167,52 @@ public CategorySync(@Nonnull final CategorySyncOptions syncOptions) {
...
+                        prepareDraftsForProcessing(categoryDrafts, keyToIdCache);
+                        categoryKeysToFetch =
+                                existingCategoryDrafts.stream().map(CategoryDraft::getKey).collect(Collectors.toSet());
+                        return categoryService
+                                .createCategories(newCategoryDrafts)
+                                .thenAccept(this::processCreatedCategories)
+                                .thenCompose(result -> categoryService
+                                        .fetchMatchingCategoriesByKeys(categoryKeysToFetch)
+                                        .handle(ImmutablePair::new)
+                                        .thenCompose(fetchResponse -> {
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: 1a7e92c

File: src/main/java/com/commercetools/sync/categories/CategorySync.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: https://github.com/commercetools/commercetools-sync-java/issues/343 opened an issue

**Comment by heshamMassoud** [comment]
> I've just tried to implement it as an extra class and this seems not good(looks like extra wrapping). But maybe we can write this functions to BaseService like below but I realized, it's not a big improvement. 

@ahmetoz ok, I really like the idea of the `isCached` method. Because we can change the logic for caching from within the method itself and not affect all the services which are using them. I will keep this into consideration on while working on reference resolution PR since it's related to the caching logic 😉 thanks for the suggestion.

**Pull request merged**